src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs

```diff
@@ -61,8 +63,15 @@ public void Execute(GeneratorExecutionContext context)
                 if (diagnostics.IsEmpty)
                 {
                     // Overwrite the cached result with the new result. This is an opportunistic cache, so as long as
-                    // the write is atomic (which it is for SetTarget) synchronization is unnecessary.
-                    s_cachedResult.SetTarget(new CachedSourceGeneratorResult(currentChecksum, sources));
+                    // the write is atomic (which it is for SetTarget) synchronization is unnecessary. We allocate an
+                    // array on the Large Object Heap (which is always part of Generation 2) and give it a reference to
+                    // the cached object to ensure this weak reference is not reclaimed prior to a full GC pass.
+                    var result = new CachedSourceGeneratorResult(currentChecksum, sources);
+                    var largeArray = new CachedSourceGeneratorResult[85000 / Unsafe.SizeOf<CachedSourceGeneratorResult>()];
```

Comment:
Just FYI, the LOH threshold is configurable.

Reply:
It is, but the threshold isn't detectable and VS uses the default. This also came up here:
https://github.com/dotnet/roslyn/blob/2f607bdd522ff206bae7883410ed52a35581a10c/src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs#L77-L79

Reply:
Everyone hard codes this number https://github.com/dotnet/runtime/issues/37693.